drv/auxflash-server/src/main.rs

```diff
@@ -0,0 +1,313 @@
+        }
+        Ok(())
+    }
+}
+
+impl idl::InOrderAuxFlashImpl for ServerImpl {
+    fn read_id(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<AuxFlashId, RequestError<AuxFlashError>> {
+        let mut idbuf = [0; 20];
+        self.qspi.read_id(&mut idbuf);
+        Ok(AuxFlashId(idbuf))
+    }
+
+    fn read_status(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError<AuxFlashError>> {
+        Ok(self.qspi.read_status())
+    }
+
+    fn slot_count(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u32, RequestError<AuxFlashError>> {
+        Ok(SLOT_COUNT)
+    }
+
+    fn read_slot_chck(
+        &mut self,
+        _: &RecvMessage,
+        slot: u32,
+    ) -> Result<AuxFlashChecksum, RequestError<AuxFlashError>> {
+        if slot >= SLOT_COUNT {
+            return Err(AuxFlashError::InvalidSlot.into());
+        }
+        let handle = SlotReader {
+            qspi: &self.qspi,
+            base: slot * 1024 * 1024,
+        };
+        let mut reader = TlvcReader::begin(handle)
+            .map_err(|_| AuxFlashError::TlvcReaderBeginFailed)?;
+
+        let mut chck_expected = None;
+        let mut chck_actual = None;
+        while let Ok(Some(chunk)) = reader.next() {
+            if &chunk.header().tag == b"CHCK" {
+                if chck_expected.is_some() {
+                    return Err(AuxFlashError::MultipleChck.into());
+                } else if chunk.len() != 32 {
+                    return Err(AuxFlashError::BadChckSize.into());
+                }
+                let mut out = [0; 32];
+                chunk
+                    .read_exact(0, &mut out)
+                    .map_err(|_| AuxFlashError::ChunkReadFail)?;
+                chck_expected = Some(out);
+            } else if &chunk.header().tag == b"AUXI" {
+                let mut sha = Sha3_256::new();
+                let mut scratch = [0u8; 256];
+                let mut i: u64 = 0;
+                while i < chunk.len() {
+                    let amount = (chunk.len() - i).min(scratch.len() as u64);
+                    chunk
+                        .read_exact(i, &mut scratch[0..(amount as usize)])
+                        .map_err(|_| AuxFlashError::ChunkReadFail)?;
+                    i += amount as u64;
+                    sha.update(&scratch[0..(amount as usize)]);
+                }
+                let mut out = [0; 32];
+                let sha_out = sha.finalize();
+                out.copy_from_slice(sha_out.as_slice());
+                chck_actual = Some(out);
+            }
+        }
+        if chck_expected.is_none() {
+            return Err(AuxFlashError::MissingChck.into());
+        } else if chck_actual.is_none() {
+            return Err(AuxFlashError::MissingAuxi.into());
+        } else if chck_expected != chck_actual {
+            return Err(AuxFlashError::ChckMismatch.into());
+        } else {
+            return Ok(AuxFlashChecksum(chck_expected.unwrap()));
```

Comment:
I might write this as a `match` just to avoid the unwrap, something like:

```rust
match (chck_expected, chck_actual) {
    (None, _) => Err(AuxFlashError::MissingChck.into()),
    (_, None) => Err(AuxFlashError::MissingAuxi.into()),
    (Some(expected), Some(actual)) if expected == actual => Ok(AuxFlashChecksum(expected)),
    (Some(_), Some(_)) => Err(AuxFlashError::ChckMismatch.into()),
}
```

but not sure if it's really better :man_shrugging:

Reply:
Good idea, I ended up with
```
        match (chck_expected, chck_actual) {
            (None, _) => Err(AuxFlashError::MissingChck.into()),
            (_, None) => Err(AuxFlashError::MissingChck.into()),
            (Some(a), Some(b)) => {
                if a == b {
                    Err(AuxFlashError::ChckMismatch.into())
                } else {
                    Ok(AuxFlashChecksum(chck_expected.unwrap()))
                }
            }
        }
```
which is a pleasant middle ground